packages/api/models/executions.js

```diff
@@ -21,33 +24,50 @@ class Execution extends Manager {
     });
   }
 
-  generateDocFromPayload(payload) {
-    const name = get(payload, 'cumulus_meta.execution_name');
+  /**
+   * Generate an execution record from a workflow execution message.
+   *
+   * @param {Object} message - A workflow execution message
+   * @returns {Object} An execution record
+   */
+  static async generateRecord(message) {
+    const executionName = getMessageExecutionName(message);
+    const stateMachineArn = getMessageStateMachineArn(message);
     const arn = getExecutionArn(
-      get(payload, 'cumulus_meta.state_machine'),
-      name
+      stateMachineArn,
+      executionName
     );
-    if (!arn) {
-      throw new Error('State Machine Arn is missing. Must be included in the cumulus_meta');
-    }
 
     const execution = aws.getExecutionUrl(arn);
-    const collectionId = getCollectionIdFromMessage(payload);
+    const collectionId = getCollectionIdFromMessage(message);
 
-    const doc = {
-      name,
+    const status = get(message, 'meta.status', 'unknown');
+
+    const record = {
+      name: executionName,
       arn,
-      parentArn: get(payload, 'cumulus_meta.parentExecutionArn'),
+      parentArn: get(message, 'cumulus_meta.parentExecutionArn'),
       execution,
-      tasks: get(payload, 'meta.workflow_tasks'),
-      error: parseException(payload.exception),
-      type: get(payload, 'meta.workflow_name'),
+      tasks: get(message, 'meta.workflow_tasks'),
+      error: parseException(message.exception),
+      type: get(message, 'meta.workflow_name'),
       collectionId,
-      status: get(payload, 'meta.status', 'unknown'),
-      createdAt: get(payload, 'cumulus_meta.workflow_start_time'),
+      status,
+      createdAt: get(message, 'cumulus_meta.workflow_start_time'),
       timestamp: Date.now()
     };
-    return doc;
+
+    const currentPayload = get(message, 'payload');
+    if (['failed', 'completed'].includes(status)) {
```

Comment:
Good question. As part of the integration with workflows, I don't think so? Because the workflow start status step should definitely only run once when the workflow starts.

And when this transitions to be triggered by Cloudwatch events, I think we'd be trusting Cloudwatch events not to send multiple events for the Step Function entering the "running" state, which I believe is how it is designed. 

Does that sound right?

Reply:
Also, I think that concern existed even with the previous implementation?

https://github.com/nasa/cumulus/pull/1180/files/28a51745b2a3989176d5b543485db354e832d44e#diff-84ba57dba2e992ae8f62c00c7af4dcceL119

`createExecutionFromSns` would be called anytime an execution was not `failed` or `completed`, so multiple "running" messages would have overridden the originalPayload

Reply:
Argh, my previous comment got deleted. 

For the current implementation, I think having the trigger for the publish lambda being a workflow step guarantees it only runs once for a given execution?

And for the upcoming implementation to use a cloudwatch event trigger, I believe it is designed to only send one update when the execution transitions to a "running" state, so we're relying on AWS to guarantee that. 

Does that make sense?

Reply:
Yeah it does and I think it's fine. I just wanted to bring it up in case anyone could think of a reason we'd need to check.